hazelcast/src/main/java/com/hazelcast/jet/impl/JobClassLoaderService.java

```diff
@@ -0,0 +1,323 @@
+                return jobClassLoaders;
+            }
+        });
+    }
+
+    /**
+     * Returns the job classloader for the job with given id.
+     *
+     * @param jobId job id
+     * @return the job classloader, null if the classloader hasn't been created yet or was already destroyed
+     */
+    public JetDelegatingClassLoader getClassLoader(long jobId) {
+        JobClassLoaders jobClassLoaders = classLoaders.get(jobId);
+        return jobClassLoaders == null ? null : jobClassLoaders.jobClassLoader;
+    }
+
+    /**
+     * Phase where the classloader is needed
+     */
+    public enum JobPhase {
+
+        /**
+         * Needed on job coordinator (master) for Pipeline/DAG deserialization and MetaSupplier init/get/close
+         */
+        COORDINATOR,
+
+        /**
+         * Needed on member for ExecutionPlan deserialization, ProcessorSupplier init/get/close and Processors
+         */
+        EXECUTION
+    }
+
+    /**
+     * Keeps job classloader and potentially processor classloaders for a job.
+     *
+     * Note:
+     * On master node there is a race between closing PMS and PS.
+     * We need to close the classloader only after both have been called.
+     * When we create JobClassLoaders we record that by calling {@link #recordPhase(JobPhase)} and
+     * when we release it we call {@link #removePhase(JobPhase)}
+     */
+    private static class JobClassLoaders {
+
+        private final JetDelegatingClassLoader jobClassLoader;
+        private final Map<String, ClassLoader> processorCls;
+
+        // EnumSet is not thread-safe
+        // Interactions are synchronized on this. Its reference never leaks the JobClassLoaders instance.
+        private final EnumSet<JobPhase> phases = EnumSet.noneOf(JobPhase.class);
+
+        JobClassLoaders(
+                @Nonnull JetDelegatingClassLoader jobClassLoader,
+                @Nonnull Map<String, ClassLoader> processorCls
+        ) {
+            this.jobClassLoader = jobClassLoader;
+            this.processorCls = unmodifiableMap(processorCls);
+        }
+
+        public JetDelegatingClassLoader jobClassLoader() {
+            return jobClassLoader;
+        }
+
+        public Map<String, ClassLoader> processorCls() {
+            return processorCls;
+        }
+
+        public ClassLoader processorCl(String key) {
+            return processorCls.get(key);
+        }
+
+        public void recordPhase(JobPhase phase) {
+            synchronized (this) {
+                phases.add(phase);
+            }
+        }
+
+        public int removePhase(JobPhase phase) {
+            synchronized (this) {
+                phases.remove(phase);
+                return phases.size();
+            }
+        }
+
+        @Override
+        public String toString() {
+            synchronized (this) {
+                return "JobClassLoaders{" +
+                        "phases=" + phases +
```

Comment:
I'd rather remove `phases` from the text and avoided synchronization.

Reply:
When the check in the JetTestSupport detects a leaked classloader it is really useful to know which phase wasn't cleaned properly.